### PR TITLE
Add mTLS support to Mixer MCP stack

### PR DESCRIPTION
**Pull request by ayj**
Add mTLS support to Mixer MCP stack.

- Add a command-line flag to specify the location of the certificate files.
- Jam the parameter through the config initialization stack to the MCP backend (sorry, I am not proud of this).
- Optionally enable using secure mode (based on the URL)

(@ozevren is OOO and asked me to take over and get this PR merged)

### CODE REVIEW

**Comment by ozevren** [comment]
It creates a non trivial bootstrap problem as Envoy gets its config from
Pilot which takes from Galley.

**Comment by douglas-reid** [comment]
@ozevren the mixer envoys are specially-configured (their filters, etc., are hard-coded). they can be bootstrapped out of that loop, right?  If this is just a matter of configuring mTLS, these mechanisms exist already.  I'm not sure where the bootstrapping becomes a problem, but I don't fully know where the concern is.

**Comment by kyessenov** [comment]
Envoy does not require Pilot. @douglas-reid is right, both Mixer's and Pilot's Envoys operate independently from everything but Citadel.

How do you perform secure naming function in go's mTLS stack?

File: pkg/mcp/creds/watcher.go [hunk 1]
```diff
@@ -21,20 +21,31 @@ import (
...
+	"github.com/spf13/cobra"
 
 	"istio.io/istio/pkg/log"
 )
 
 var scope = log.RegisterScope("mcp-creds", "MCP Credential utilities", 0)
 
+const (
+	// DefaultCertDir is the default directory in which MCP options reside.
+	DefaultCertDir = "/etc/istio/certs/"
```
Comment (geeknoid) [inline pkg/mcp/creds/watcher.go]: Do these constant need to be public?

**Comment by ayj** [comment]
> Curious: couldn't we leverage mTLS support in the proxies that sit in front of Mixer? Or is MCP / Galley mTLS out-of-band somehow?

I believe ToC is planning to discuss this issue next week on whether Mixer/Pilot should continue to use sidecars or manage mTLS, etc. directly with a library. Is it reasonable to defer switching to sidecar for MCP for one week based on that decision? I'm happy to make the change (or not, either way), but don't want to spend time on something that could potentially be thrown away in a week. cc @geeknoid

File: pkg/mcp/creds/watcher.go (hunk 1 above)
Comment (ayj) [reply pkg/mcp/creds/watcher.go]: nope - unexported

**Comment by douglas-reid** [comment]
@ayj Sure. One of my main concerns is that whatever happens here would complicate the conversion to sidecar-less Mixer. And in the mean time, having two separate places configuring mTLS (one driven from mesh config control plane auth flag, the other from a binary-specific configuration) seems a bit non-ideal.

@geeknoid hopefully the ToC has data to inform any decision on switching from an envoy-based mTLS solution to a custom library that performs the same function. 


I do want to point out that there is zero chance we can throw away the Mixer sidecars within a week, whatever the decision.

Regarding throwaway work, I don't know, but I imagine the work to have Mixer's calls to Galley be over mTLS wouldn't require a large change in the container config for Mixer. I do wonder... if we want to secure communication from Mixer to gRPC adapters, it would seem like having mTLS out is a requirement. Any work to capture egress from Mixer and use mTLS where appropriate could immediately pay dividends.

**Comment by ayj** [comment]
> How do you perform secure naming function in go's mTLS stack?

Galley's cert has DNS name and secure naming isn't needed with DNS. Galley server uses `VerifyPeerCertificate` to verify client's spiffie identity using a basic access list checker. If ToC decision is to use sidecar for _all_ control plane connections we'll still need to address ACL'ing as Istio RBAC won't be fully available during bootstrapping.

File: mixer/pkg/config/mcp/backend.go [hunk 2]
```diff
@@ -115,12 +141,47 @@ func (b *backend) Init(kinds []string) error {
 	}
 
 	ctx, cancel := context.WithCancel(context.Background())
-	conn, err := grpc.DialContext(ctx, b.serverAddress, grpc.WithInsecure())
+
```
Comment (geeknoid) [inline mixer/pkg/config/mcp/backend.go]: I think all this new code needs test coverage

File: pkg/mcp/creds/watcher.go [hunk 3]
```diff
@@ -53,23 +64,54 @@ type CertificateWatcher struct {
...
+		KeyFile:           filepath.Join(defaultCertDir, defaultKeyFile),
+		CACertificateFile: filepath.Join(defaultCertDir, defaultCACertificateFile),
+	}
+}
+
+// AttachCobraFlags attaches a set of Cobra flags to the given Cobra command.
+//
+// Cobra is the command-line processor that Istio uses. This command attaches
+// the necessary set of flags to configure the MCP options.
+func (c *Options) AttachCobraFlags(cmd *cobra.Command) {
```
Comment (geeknoid) [inline pkg/mcp/creds/watcher.go]: Any code coverage for this?

**Pull request merged**